### DISCUSSION TITLE
Access assets from functions in state store

### DISCUSSION THREAD

**Discussion by ghost** [quasarframework/quasar#10265] [General - Components / Directives / etc]
I would like to access images in the `public` folder from functions in my `store`?  Is that possible?  I have had a look at [this](https://quasar.dev/quasar-cli/developing-electron-apps/electron-accessing-files#read-and-write-local-files), but it looks like it is only for accessing via Vue files.  Is it possible to access it via a nodejs local file path?

```
function makeHeader(state) {
    console.log(__dirname); // returns /
    return {
        image: 'header.png'
    }
}
```

**Comment by ontwikkelfabriek** [comment]
@nickberry17 

>  I would like to access images in the public folder from functions in my store? Is that possible?

Yes you can do that.

> I have had a look at this, but it looks like it is only for accessing via Vue files.

That's Electron specific documentation for reading and writing **local** files. For handling **static** assets in the **public** folder look here:
 https://quasar.dev/quasar-cli/handling-assets#introduction

**Reply by ghost** [reply]
Thank you for that, but I still have a couple of issues.  First, I can only get it to compile if I reference the image path using a relative path (`../../public/img/header.png`).   It compiles, but when I run the function in the browser it returns this error:

`Uncaught (in promise) Invalid image: File 'http://localhost:8080/img/header.png' not found in virtual file system`

I don't know where that `img` directory is supposed to be.  Do you know why it's doing this?  

Regarding the image path...

This is what compiles
```
function makeHeader() {
    return {
        image: require('../../public/img/header.png')
    }
}
```

This doesn't compile, but based on my understanding of the docs, this should work.
` require('./img/header.png')`
`Module not found: Can't resolve imported dependency "./img/header.png"
`

Here's the directory structure of the app
![dir](https://user-images.githubusercontent.com/18670565/128279368-a9176c56-58a0-4b45-bba8-ee8d541d4943.png)

**Reply by ontwikkelfabriek** [reply]
@nickberry17

If you create a small working codepen.io that demo's your problem, we can take a look at it.

**Reply by ghost** [reply]
Can I do one better and link you the repo?  https://github.com/nickberry17/quasarRG 
Run that then go to Create PDF tab and click the Create PDF button, no need to fill the form.

**Reply by ontwikkelfabriek** [reply]
@nickberry17 

Sure! I will take a look tomorrow.

**Reply by ontwikkelfabriek** [reply]
@nickberry17 

I took a look. 

The problem is that the pdfMake lib does not accept `http//` urls. It will only accept relative paths. So you can either:
1) use `../../public/img/header.png` like you did before
2) or use a base64 encoded image. Like they do over here with `testImageUrl`:
https://github.com/bpampuch/pdfmake/blob/0.2/examples/images.js
3) use another JS Pdf library:
https://openbase.com/categories/js/best-nodejs-pdf-generator-libraries

Another thing is that you are using `src/store/index.js` to put your custom 'store/functions' in. This location is reserved for `Vuex` so I would choose another location for your pdf functions. Something like `src/services/pdf/index.js` or `/src/services/pdfFooBar.js`

**Reply by ghost** [reply]
Hey, I really appreciate you taking the time to look into that, I'm not sure I want to switch to another PDF Library just yet because I want something that's declarative.  Maybe I need to ask on the PDFMake project, but using the relative path doesn't work.  Error is:

```
Invalid image: File '../../public/img/header.png' not found in virtual file system
Images dictionary should contain dataURL entries (or local file paths in node.js)
```

Come to think of it, I was not able to use any images in `src/assets/` in the UI (as opposed to from within PDFMake) either.  I am not using Vuex, just using this as my own store, but would that be a source of problems in this case?  I will go ahead and reorganize the functions and state store into another location, but did you mean that I need to move those to get this working?
